File: lib/domain/usecases/task/update_task_usecase.dart
import '../../entities/commit.dart' as entity;
import '../../entities/enums.dart';
import '../../entities/task.dart';
import '../../repositories/i_commit_repository.dart';
import '../../repositories/i_task_repository.dart';

/// 更新任务用例
class UpdateTaskUseCase {
  UpdateTaskUseCase(
    this._taskRepository,
    this._commitRepository,
  );

  final ITaskRepository _taskRepository;
  final ICommitRepository _commitRepository;

  Future<Task> execute({
    required String id,
    String? title,
    String? description,
    TaskStatus? status,
    Priority? priority,
    DateTime? dueDate,
    int? sortOrder,
  }) async {
    final existing = await _taskRepository.getById(id);
    if (existing == null) {
      throw ArgumentError('任务不存在: $id');
    }

    // 如果状态变更，验证状态机
    if (status != null && status != existing.status) {
      if (!TaskStateMachine.canTransition(
        existing.status,
        status,
      )) {
        throw StateError(
          '非法状态转换: ${existing.status} -> $status',
        );
      }
    }

    final updated = existing.copyWith(
      title: title ?? existing.title,
      description: description ?? existing.description,
      status: status ?? existing.status,
      priority: priority ?? existing.priority,
      dueDate: dueDate ?? existing.dueDate,
      sortOrder: sortOrder ?? existing.sortOrder,
    );

    final savedTask = await _taskRepository.update(updated);

    // 记录更新提交
    await _commitRepository.create(
      entity.Commit.create(
        taskId: savedTask.id,
        branchId: savedTask.branchId,
        message: '更新任务: ${savedTask.title}',
        type: CommitType.update,
      ),
    );

    return savedTask;
  }
}

/// 任务状态机
class TaskStateMachine {
  static bool canTransition(
    TaskStatus from,
    TaskStatus to,
  ) {
    return switch (from) {
      TaskStatus.todo => [
          TaskStatus.inProgress,
          TaskStatus.done,
          TaskStatus.cancelled,
        ].contains(to),
      TaskStatus.inProgress => [
          TaskStatus.todo,
          TaskStatus.done,
          TaskStatus.cancelled,
        ].contains(to),
      TaskStatus.done => [
          TaskStatus.todo,
        ].contains(to),
      TaskStatus.cancelled => [
          TaskStatus.todo,
        ].contains(to),
    };
  }
}

File: web/src/application/usecases/task/update-task-usecase.ts
import { Commit } from '../../../domain/entities/commit';
import { CommitType, Priority, TaskStatus } from '../../../domain/entities/enums';
import { Task } from '../../../domain/entities/task';
import { ICommitRepository } from '../../../domain/repositories/i-commit-repository';
import { ITaskRepository } from '../../../domain/repositories/i-task-repository';
import { validateNotEmpty } from '../../../core/utils/validators';

export interface UpdateTaskInput {
  id: string;
  title?: string;
  description?: string | null;
  status?: TaskStatus;
  priority?: Priority;
  dueDate?: Date | null;
  parentTaskId?: string | null;
  sortOrder?: number;
}

export class UpdateTaskUseCase {
  constructor(
    private taskRepo: ITaskRepository,
    private commitRepo: ICommitRepository
  ) {}

  async execute(input: UpdateTaskInput): Promise<Task> {
    validateNotEmpty(input.id, '任务 ID');
    const task = await this.taskRepo.getById(input.id);
    if (!task) {
      throw new Error('任务不存在');
    }
    if (input.title !== undefined) {
      validateNotEmpty(input.title, '任务标题');
    }

    const updated = task.copyWith(input);
    const result = await this.taskRepo.update(updated);

    const commit = Commit.create(
      result.id,
      result.branchId,
      `更新任务 ${result.title}`,
      CommitType.update
    );
    await this.commitRepo.create(commit);
    return result;
  }
}
